File: mip_reference/ruby-sinatra/lib/mip/models/member.rb
# frozen_string_literal: true

module MIP
  module Models
    # Represents a member in the local organization
    class Member
      attr_accessor :member_number, :prefix, :first_name, :middle_name, :last_name,
                    :suffix, :honorific, :rank, :birthdate, :years_in_good_standing,
                    :status, :is_active, :good_standing, :email, :phone, :cell,
                    :address, :affiliations, :life_cycle_events

      def initialize(attrs = {})
        @member_number = attrs[:member_number]
        @prefix = attrs[:prefix]
        @first_name = attrs[:first_name]
        @middle_name = attrs[:middle_name]
        @last_name = attrs[:last_name]
        @suffix = attrs[:suffix]
        @honorific = attrs[:honorific]
        @rank = attrs[:rank]
        @birthdate = attrs[:birthdate]
        @years_in_good_standing = attrs[:years_in_good_standing]
        @status = attrs.fetch(:status, 'Active')
        @is_active = attrs.fetch(:is_active, true)
        @good_standing = attrs.fetch(:good_standing, true)
        @email = attrs[:email]
        @phone = attrs[:phone]
        @cell = attrs[:cell]
        @address = attrs[:address] || {}
        @affiliations = attrs[:affiliations] || []
        @life_cycle_events = attrs[:life_cycle_events] || []
      end

      def full_name
        [
          @prefix,
          @first_name,
          @middle_name,
          @last_name,
          @suffix
        ].compact.reject(&:empty?).join(' ')
      end

      def party_short_name
        "#{@first_name} #{@last_name[0]}"
      end

      # Member type from first active affiliation
      def member_type
        active_affiliation = @affiliations.find { |a| a[:is_active] }
        active_affiliation&.dig(:member_type) || 'Member'
      end

      # Format for search response
      def to_search_result
        {
          member_number: @member_number,
          first_name: @first_name,
          last_name: @last_name,
          birthdate: @birthdate,
          contact: {
            email: @email,
            phone: @phone,
            address: @address
          },
          group_status: {
            status: @status,
            is_active: @is_active,
            good_standing: @good_standing
          },
          affiliations: @affiliations.map do |aff|
            {
              local_name: aff[:local_name],
              local_status: aff[:local_status] || aff[:status],
              is_active: aff[:is_active],
              member_type: aff[:member_type]
            }
          end
        }
      end

      # Full member profile for COGS
      def to_member_profile
        {
          member_number: @member_number,
          prefix: @prefix,
          first_name: @first_name,
          middle_name: @middle_name,
          last_name: @last_name,
          suffix: @suffix,
          honorific: @honorific,
          rank: @rank,
          birthdate: @birthdate,
          years_in_good_standing: @years_in_good_standing,
          group_status: {
            status: @status,
            is_active: @is_active
          },
          contact: {
            email: @email,
            phone: @phone,
            cell: @cell,
            address: @address
          },
          affiliations: @affiliations,
          life_cycle_events: @life_cycle_events
        }
      end

      # Quick status check response
      def to_status_check
        {
          member_number: @member_number,
          member_type: member_type,
          party_short_name: party_short_name,
          group_status: {
            status: @status,
            is_active: @is_active,
            good_standing: @good_standing
          }
        }
      end

      # Create from config YAML
      def self.from_config(config)
        new(
          member_number: config['member_number'],
          prefix: config['prefix'],
          first_name: config['first_name'],
          middle_name: config['middle_name'],
          last_name: config['last_name'],
          suffix: config['suffix'],
          honorific: config['honorific'],
          rank: config['rank'],
          birthdate: config['birthdate'],
          years_in_good_standing: config['years_in_good_standing'],
          status: config.fetch('status', 'Active'),
          is_active: config.fetch('is_active', true),
          good_standing: config.fetch('good_standing', true),
          email: config['email'],
          phone: config['phone'],
          cell: config['cell'],
          address: symbolize_keys(config['address'] || {}),
          affiliations: (config['affiliations'] || []).map { |a| symbolize_keys(a) },
          life_cycle_events: (config['life_cycle_events'] || []).map { |e| symbolize_keys(e) }
        )
      end

      def self.symbolize_keys(hash)
        hash.transform_keys(&:to_sym)
      end
    end
  end
end

File: mip_reference/node-express/lib/mip/models/member.js
'use strict';

class Member {
  constructor(attrs = {}) {
    this.memberNumber = attrs.memberNumber;
    this.prefix = attrs.prefix || null;
    this.firstName = attrs.firstName;
    this.middleName = attrs.middleName || null;
    this.lastName = attrs.lastName;
    this.suffix = attrs.suffix || null;
    this.honorific = attrs.honorific || null;
    this.rank = attrs.rank || null;
    this.birthdate = attrs.birthdate || null;
    this.yearsInGoodStanding = attrs.yearsInGoodStanding || 0;
    this.status = attrs.status || 'Active';
    this.isActive = attrs.isActive !== undefined ? attrs.isActive : true;
    this.goodStanding = attrs.goodStanding !== undefined ? attrs.goodStanding : true;
    this.email = attrs.email || null;
    this.phone = attrs.phone || null;
    this.cell = attrs.cell || null;
    this.address = attrs.address || {};
    this.affiliations = attrs.affiliations || [];
    this.lifeCycleEvents = attrs.lifeCycleEvents || [];
  }

  fullName() {
    return [this.prefix, this.firstName, this.middleName, this.lastName, this.suffix]
      .filter(x => x && x.length > 0)
      .join(' ');
  }

  partyShortName() {
    return `${this.firstName} ${this.lastName[0]}`;
  }

  // Member type from first active affiliation
  memberType() {
    const active = this.affiliations.find(a => a.is_active);
    return (active && active.member_type) || 'Member';
  }

  // Format for search response
  toSearchResult() {
    return {
      member_number: this.memberNumber,
      first_name: this.firstName,
      last_name: this.lastName,
      birthdate: this.birthdate,
      contact: {
        email: this.email,
        phone: this.phone,
        address: this.address
      },
      group_status: {
        status: this.status,
        is_active: this.isActive,
        good_standing: this.goodStanding
      },
      affiliations: this.affiliations.map(aff => ({
        local_name: aff.local_name,
        local_status: aff.local_status || aff.status,
        is_active: aff.is_active,
        member_type: aff.member_type
      }))
    };
  }

  // Full member profile for COGS
  toMemberProfile() {
    return {
      member_number: this.memberNumber,
      prefix: this.prefix,
      first_name: this.firstName,
      middle_name: this.middleName,
      last_name: this.lastName,
      suffix: this.suffix,
      honorific: this.honorific,
      rank: this.rank,
      birthdate: this.birthdate,
      years_in_good_standing: this.yearsInGoodStanding,
      group_status: {
        status: this.status,
        is_active: this.isActive
      },
      contact: {
        email: this.email,
        phone: this.phone,
        cell: this.cell,
        address: this.address
      },
      affiliations: this.affiliations,
      life_cycle_events: this.lifeCycleEvents
    };
  }

  // Quick status check response
  toStatusCheck() {
    return {
      member_number: this.memberNumber,
      member_type: this.memberType(),
      party_short_name: this.partyShortName(),
      group_status: {
        status: this.status,
        is_active: this.isActive,
        good_standing: this.goodStanding
      }
    };
  }

  // Create from config YAML
  static fromConfig(config) {
    return new Member({
      memberNumber: config.member_number,
      prefix: config.prefix,
      firstName: config.first_name,
      middleName: config.middle_name,
      lastName: config.last_name,
      suffix: config.suffix,
      honorific: config.honorific,
      rank: config.rank,
      birthdate: config.birthdate,
      yearsInGoodStanding: config.years_in_good_standing,
      status: config.status || 'Active',
      isActive: config.is_active !== undefined ? config.is_active : true,
      goodStanding: config.good_standing !== undefined ? config.good_standing : true,
      email: config.email,
      phone: config.phone,
      cell: config.cell,
      address: config.address || {},
      affiliations: config.affiliations || [],
      lifeCycleEvents: config.life_cycle_events || []
    });
  }
}

module.exports = Member;
